rust/xaynet-server/src/services/messages/multipart.rs

```diff
@@ -0,0 +1,474 @@
+use std::{
+    collections::{BTreeMap, HashMap},
+    task::Poll,
+};
+
+use futures::{
+    future::{self, Ready},
+    task::Context,
+};
+use tower::Service;
+use xaynet_core::{
+    crypto::{PublicEncryptKey, PublicSigningKey},
+    message::{Chunk, DecodeError, FromBytes, Message, Payload, Sum, Sum2, Tag, Update},
+};
+
+use crate::services::messages::ServiceError;
+
+/// A `PartialMessage` stores chunks of a multipart message. Once it
+/// has all the chunks, it can be consumed and turned into a
+/// full-blown [`Message`] (see [`into_message`]).
+#[derive(Debug)]
+#[cfg_attr(test, derive(Clone))]
+pub struct PartialMessage {
+    /// Public key of the participant sending the message
+    participant_pk: PublicSigningKey,
+    /// Public key of the coordinator
+    coordinator_pk: PublicEncryptKey,
+    /// Message type
+    tag: Tag,
+    /// The ID of the last chunk is actually the total number of
+    /// chunks this message is made of.
+    last_chunk_id: Option<u16>,
+    /// Chunks, ordered by ID
+    data: BTreeMap<u16, Vec<u8>>,
+}
+
+impl PartialMessage {
+    /// Create a new [`PartialMessage`] that contains no chunk.
+    fn new(tag: Tag, participant_pk: PublicSigningKey, coordinator_pk: PublicEncryptKey) -> Self {
+        PartialMessage {
+            tag,
+            participant_pk,
+            coordinator_pk,
+            data: BTreeMap::new(),
+            last_chunk_id: None,
+        }
+    }
+    /// Return `true` if the message is complete, _i.e._ if all
+    /// `MultiparMessage` holds all the chunks.
+    fn has_all_chunks(&self) -> bool {
+        self.last_chunk_id
+            .map(|expected_number_of_chunks| self.data.len() == expected_number_of_chunks as usize)
+            .unwrap_or(false)
+    }
+
+    /// Return the first missing chunk. If no chunk is missing, return `None`.
+    #[allow(dead_code)]
+    fn first_missing_chunk(&self) -> Option<u16> {
+        if self
+            .last_chunk_id
+            .map(|last_id| self.data.len() >= last_id as usize)
+            .unwrap_or(false)
+        {
+            return None;
+        }
+
+        let mut current_id = 0;
+        for id in self.data.keys() {
+            if id - current_id > 1 {
+                break;
+            }
+            current_id = *id;
```

Comment:
I'm not sure if it's safer, to me it looks completely equivalent. There cannot be repeated IDs because the IDs are keys in a hashmap, so all distinct.

Reply:
I removed the method for now so I'll mark this as resolved